### PR TITLE
ISPN-5389 Optimize overlapping range and interval predicates (non-ind…

### PR DESCRIPTION
**Pull request by anistor**
…exed query)

This is for master.

jira: https://issues.jboss.org/browse/ISPN-5389

### CODE REVIEW

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/ProtobufPropertyHelper.java [hunk 1]
```diff
@@ -126,6 +134,32 @@ public boolean hasEmbeddedProperty(String entityType, List<String> propertyPath)
...
+      int i = 0;
+      for (String p : propertyPath) {
+         FieldDescriptor field = messageDescriptor.findFieldByName(p);
+         if (field == null) {
+            break;
+         }
+         if (field.isRepeated()) {
+            return true;
+         }
+         if (++i == propertyPath.size() || field.getJavaType() != JavaType.MESSAGE) {
```
Comment (danberindei) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/ProtobufPropertyHelper.java]: Do you really need the `++i == propertyPath.size()` condition here, if you're already iterating on the propertyPath?

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterConjunctionPredicate.java [hunk 2]
```diff
@@ -16,6 +16,7 @@ public FilterConjunctionPredicate() {
    @Override
    public BooleanExpr getQuery() {
       // we always expect children.size() >= 1
+      assert children.size() >= 1;
```
Comment (danberindei) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterConjunctionPredicate.java]: Is an `AssertionError` that much better than an `IndexOutOfBoundsException`?

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/predicateindex/be/PredicateNode.java [hunk 3]
```diff
@@ -57,7 +57,7 @@ public void handleChildValue(BENode child, boolean childValue, FilterEvalContext
 
       if (isEvaluationComplete(evalContext)) {
          if (predicate.isRepeated() && evalContext.treeCounters[startIndex] == value) {
-            // receiving the same value multiple times if fine if this is a repeated condition
+            // receiving the same value multiple times is fine if this is a repeated condition  //todo ?????
```
Comment (danberindei) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/predicateindex/be/PredicateNode.java]: TODO message isn't very clear :)

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/BooleanFilterNormalizer.java [hunk 4]
```diff
@@ -176,6 +192,134 @@ private int comparePrimaryPredicateExpr(boolean isFirstNegated, PrimaryPredicate
          return -1;
       }
 
+      private void optimizeOverlappingIntervalPredicates(List<BooleanExpr> children, boolean isConjunction) {
+         for (int i = 0; i < children.size(); i++) {
+            BooleanExpr ci = children.get(i);
+            if (ci instanceof ComparisonExpr) {
+               ComparisonExpr first = (ComparisonExpr) ci;
+               assert first.getLeftChild() instanceof PropertyValueExpr;
+               assert first.getRightChild() instanceof ConstantValueExpr;
```
Comment (danberindei) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/BooleanFilterNormalizer.java]: Hmmm, so you don't support a condition like `field = x or 1 = 1`? I remember using this trick a lot with SQL parametrized statements, people might use it for query as well.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/ComparisonExpr.java [hunk 5]
```diff
@@ -86,22 +88,18 @@ public boolean equals(Object o) {
...
    }
 
    @Override
    public String toString() {
-      return "ComparisonExpr{" +
-            " comparisonType=" + comparisonType +
-            ", leftChild=" + leftChild +
-            ", rightChild=" + rightChild +
-            '}';
+      return "COMP(" + leftChild + ", " + type + ", " + rightChild + ')';
```
Comment (danberindei) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/ComparisonExpr.java]: Why not `type(left, right)`?

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/BooleanFilterNormalizer.java (hunk 4 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/BooleanFilterNormalizer.java]: Comparing field with field or constant with constant is not supported. But that type of jpa query cannot be built by the DSL builder, so we are safe.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterConjunctionPredicate.java (hunk 2 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterConjunctionPredicate.java]: That assert was only meant for documentation. I think I'll turn it into a comment.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/ComparisonExpr.java (hunk 5 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/ComparisonExpr.java]: Not a bad idea :)

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/predicateindex/be/PredicateNode.java (hunk 3 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/predicateindex/be/PredicateNode.java]: TBH I do not remember now what I meant with that comment, I just fixed the spelling and then wondered if there was something to be fixed there. I'll remove the TODO.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/ProtobufPropertyHelper.java (hunk 1 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/ProtobufPropertyHelper.java]: That seems to be a leftover. Will remove it.

**Comment by anistor** [comment]
@danberindei, I've just updated this.

**Comment by danberindei** [comment]
Integrated, thanks Adrian!

**Pull request closed without merging**